node/core/approval-voting/src/lib.rs

```diff
@@ -1997,35 +2078,60 @@ async fn launch_approval(
 			.with_stage(jaeger::Stage::ApprovalChecking);
 
 		let available_data = match a_rx.await {
-			Err(_) => return,
+			Err(e) => return ApprovalState::failed(
+				validator_index,
+				candidate_hash,
+				candidate_index as _,
+				InvalidCandidate::ExecutionError(e.to_string())
+			),
 			Ok(Ok(a)) => a,
-			Ok(Err(RecoveryError::Unavailable)) => {
-				tracing::warn!(
-					target: LOG_TARGET,
-					"Data unavailable for candidate {:?}",
-					(candidate_hash, candidate.descriptor.para_id),
-				);
-				// do nothing. we'll just be a no-show and that'll cause others to rise up.
-				metrics_guard.take().on_approval_unavailable();
-				return;
-			}
-			Ok(Err(RecoveryError::Invalid)) => {
-				tracing::warn!(
-					target: LOG_TARGET,
-					"Data recovery invalid for candidate {:?}",
-					(candidate_hash, candidate.descriptor.para_id),
-				);
+			Ok(Err(e)) => {
+				match &e {
+					&RecoveryError::Unavailable => {
+						tracing::warn!(
+							target: LOG_TARGET,
+							"Data unavailable for candidate {:?}",
+							(candidate_hash, candidate.descriptor.para_id),
+						);
+						// do nothing. we'll just be a no-show and that'll cause others to rise up.
+						metrics_guard.take().on_approval_unavailable();
+					}
+					&RecoveryError::Invalid => {
+						tracing::warn!(
+							target: LOG_TARGET,
+							"Data recovery invalid for candidate {:?}",
+							(candidate_hash, candidate.descriptor.para_id),
+						);
 
-				// TODO: dispute. Either the merkle trie is bad or the erasure root is.
-				// https://github.com/paritytech/polkadot/issues/2176
-				metrics_guard.take().on_approval_invalid();
-				return;
+						// TODO: dispute. Either the merkle trie is bad or the erasure root is.
+						// https://github.com/paritytech/polkadot/issues/2176
+						metrics_guard.take().on_approval_invalid();
+					}
+				}
+				return ApprovalState::failed(
+					validator_index,
+					candidate_hash,
+					candidate_index as _,
+					InvalidCandidate::ExecutionError(e.to_string())
```

Comment:
This indicates to me that `InvalidCandidate` is not the right type. Data recovery errors are not execution errors. I don't see any real reason to propagate the `InvalidCandidate` upwards to the `CurrentlyCheckingSet`.